File: leetcode.py
import urllib2
import StringIO
import string

root = urllib2.urlopen("https://leetcode.com/problemset/algorithms/").read()
prefix = 'https://leetcode.com'

#with open("leetcode.html", "w") as f:
#	f.write(root) 

buf = StringIO.StringIO(root)

golang_count = 0

with open("leetcode_golang_support.csv", "w") as f:
	f.write("#,URL,Title,Acceptance,Difficulty,Golang Support\n");

	lines = buf.readlines()
	for idx, line_0 in enumerate(lines):
		isSubstring = "<a href=\"/problems/" in line_0
		if isSubstring:
			number = lines[idx-2].strip()
			number = string.replace(number, '<td>', '');
			number = string.replace(number, '</td>', '');
			#print number

			extra_line_number = 0
			isExtraLines = "fa fa-lock" in lines[idx+2]
			if isExtraLines:
				extra_line_number = 2

			acceptance = lines[idx+3+extra_line_number].strip()
			acceptance = string.replace(acceptance, '<td>', '');
			acceptance = string.replace(acceptance, '</td>', '');
			#print acceptance

			difficulty = lines[idx+5+extra_line_number].strip()
			difficulty = string.replace(difficulty, '</td>', '');
			i = difficulty.index('>')
			difficulty = string.replace(difficulty, difficulty[0:i+1], '');
			#print difficulty

			line = line_0.strip()
			line = string.replace(line, ',', '') #remove ',' between pow(x,n)
			line = string.replace(line, '<a href=\"', prefix)
			line = string.replace(line, '\">', ',')
			line = string.replace(line, '</a>', '')
			split_fields = line.split(',')
			#print split_fields

			golang_exist = ''
			problem = urllib2.urlopen(split_fields[0]).read()
			buf2 = StringIO.StringIO(problem)
			for line2 in buf2.readlines():
				isSubstring2 = "golang" in line2
				if isSubstring2:
					golang_exist = 'golang'
					golang_count += 1
					#print "golang"
					break

			print number+","+line+","+acceptance+","+difficulty+","+golang_exist+"\n"
			f.write(number+","+line+","+acceptance+","+difficulty+","+golang_exist+"\n")	

print "Total golang count is: "+str(golang_count)

File: leetcode.go
package main

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"sort"
	"strconv"
	"strings"
)

type langSupport struct {
	number     int
	line       string
	lang       string
	difficulty string
	locked     string
}

type langSupportSlice []langSupport

func (l langSupportSlice) Len() int           { return len(l) }
func (l langSupportSlice) Less(i, j int) bool { return l[i].number < l[j].number }
func (l langSupportSlice) Swap(i, j int)      { l[i], l[j] = l[j], l[i] }

func findLang(out chan<- langSupport, lines []string, idx int, lang string) (err error) {
	prefix := "https://leetcode.com"

	var ls langSupport

	number := strings.TrimSpace(lines[idx-2])
	number = strings.Replace(number, "<td>", "", -1)
	number = strings.Replace(number, "</td>", "", -1)
	ls.number, err = strconv.Atoi(number)
	if err != nil {
		fmt.Fprintf(os.Stderr, "parse number error: %v\n", err)
		return err
	}

	extraLineNumber := 0
	ls.locked = ""
	isExtraLines := strings.Contains(lines[idx+2], "fa fa-lock")
	if isExtraLines {
		extraLineNumber = 2
		ls.locked = "Locked"
	}

	isEditorial := strings.Contains(lines[idx+6+extraLineNumber], "true")
	if isEditorial {
		extraLineNumber += 9
	}

	ls.difficulty = strings.TrimSpace(lines[idx+10+extraLineNumber])
	ls.difficulty = strings.Replace(ls.difficulty, "</td>", "", -1)
	i := strings.Index(ls.difficulty, ">")
	ls.difficulty = strings.Replace(ls.difficulty, ls.difficulty[0:i+1], "", -1)

	ls.line = strings.TrimSpace(lines[idx])
	ls.line = strings.Replace(ls.line, ",", "", -1) //remove ',' between pow(x,n)
	ls.line = strings.Replace(ls.line, "<a href=\"", prefix, -1)
	ls.line = strings.Replace(ls.line, "\">", ",", 1)
	ls.line = strings.Replace(ls.line, "</a>", "", -1)
	splitFields := strings.Split(ls.line, ",")

	//////////////////////////////////////////
	resp2, err := http.Get(splitFields[0])
	if err != nil {
		fmt.Fprintf(os.Stderr, "fetch %s: %v\n", splitFields[0], err)
		return err
	}
	problem, err := ioutil.ReadAll(resp2.Body)
	resp2.Body.Close()
	if err != nil {
		fmt.Fprintf(os.Stderr, "fetch: reading %s: %v\n", splitFields[0], err)
		return err
	}

	ls.lang = ""
	buf2 := strings.Split(string(problem), "\n")
	for _, line2 := range buf2 {
		isSubstring2 := strings.Contains(line2, lang)
		if isSubstring2 {
			ls.lang = lang
			break
		}
	}

	out <- ls

	return nil
}

func main() {
	if len(os.Args) != 2 {
		println("Usage: leetcode language")
		println("\t language: C/C++/golang/swift...")
		return
	}

	lang := os.Args[1]
	url := "https://leetcode.com/problemset/algorithms/"

	resp, err := http.Get(url)
	if err != nil {
		fmt.Fprintf(os.Stderr, "fetch: %v\n", err)
		return
	}
	root, err := ioutil.ReadAll(resp.Body)
	resp.Body.Close()
	if err != nil {
		fmt.Fprintf(os.Stderr, "fetch: reading %s: %v\n", url, err)
		return
	}

	/*fp_html, err := os.Create("leetcode.html")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Create file leetcode.html failed\n")
		return
	}
	defer fp_html.Close()
	fp_html.Write(root)*/

	fpLang, err := os.Create("leetcode_" + lang + "_support.csv")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Create file leetcode_lang_support.csv failed\n")
		return
	}
	defer fpLang.Close()

	fpLang.WriteString("#,URL,Title,Support,Difficulty,Locked\n")

	langSupportChan := make(chan langSupport, 8)
	count := 0
	lines := strings.Split(string(root), "\n")
	for idx, line := range lines {
		isSubstring := strings.Contains(line, "<a href=\"/problems/")
		if isSubstring {
			count++
			go findLang(langSupportChan, lines, idx, lang)
		}
	}

	lss := make(langSupportSlice, count)
	for i := 0; i < count; i++ {
		lss[i] = <-langSupportChan
		print(strconv.Itoa(lss[i].number) + "," + lss[i].line + "," + lss[i].lang + "," + lss[i].difficulty + "," + lss[i].locked + "\n")
	}

	sort.Sort(lss)

	for i := 0; i < count; i++ {
		fpLang.WriteString(strconv.Itoa(lss[i].number) + "," + lss[i].line + "," + lss[i].lang + "," + lss[i].difficulty + "," + lss[i].locked + "\n")
	}
}
